acceptance/tests/provider/package/gem.rb

```diff
@@ -0,0 +1,96 @@
+    #
+    # Magically, the puppet command succeeds on a Linux host with both the 'master' and 'agent' roles.
+    next unless agent['roles'].include?('master') || agent['template'].include?('windows')
+
+    beaker_path = agent.get_env_var('PATH')
+
+    step "Setup: Install Ruby, and reset PATH on Linux" do
+      if agent['platform'].include?('windows')
+        on(agent, powershell(wins_download_ruby), :catch_failures => true)
+        on(agent, powershell(wins_install_ruby), :catch_failures => true)
+      else
+        agent.add_env_var('PATH', puppet_agent_sh_path)
+        package_present(agent, 'ruby')
+      end
+    end
+
+    step "Install a gem package" do
+      package_manifest = resource_manifest('package', package, { ensure: 'present', provider: 'gem' } )
+      apply_manifest_on(agent, package_manifest, :catch_failures => true) do
+        list = on(agent, 'gem list').stdout
+        assert_match(/#{package} \(/, list)
+      end
+      on(agent, "gem uninstall #{package}")
+    end
+
+    step "Uninstall a gem package" do
+      on(agent, "gem install #{package}")
+      package_manifest = resource_manifest('package', package, { ensure: 'absent', provider: 'gem' } )
+      apply_manifest_on(agent, package_manifest, :catch_failures => true) do
+        list = on(agent, 'gem list').stdout
+        assert_no_match(/#{package} \(/, list)
+      end
+    end
+
+    # Puppet's Ruby is a convenient Ruby that should not be first in the PATH.
+    # Another could be: https://www.softwarecollections.org/en/scls/rhscl/rh-ruby23/
+
+    puppet_gem_command = "#{agent['privatebindir']}#{File::SEPARATOR}gem"
+
+    step "Install a gem package with a target command" do
+      package_manifest = resource_manifest('package', package, { ensure: 'present', provider: 'gem', command: puppet_gem_command } )
+      apply_manifest_on(agent, package_manifest, :catch_failures => true) do
+        list = on(agent, "#{puppet_gem_command} list").stdout
+        assert_match(/#{package} \(/, list)
+      end
+      on(agent, "#{puppet_gem_command} uninstall #{package}")
+    end
+
+    step "Uninstall a gem package with a target command" do
+      on(agent, "#{puppet_gem_command} install #{package}")
+      package_manifest = resource_manifest('package', package, { ensure: 'absent', provider: 'gem', command: puppet_gem_command } )
+      apply_manifest_on(agent, package_manifest, :catch_failures => true) do
+        list = on(agent, "#{puppet_gem_command} list").stdout
+        assert_no_match(/#{package} \(/, list)
+      end
+      on(agent, "#{puppet_gem_command} uninstall #{package}")
+    end
+
+    step "Teardown: Remove Ruby, and reset PATH on Linux" do
+      if agent['platform'].include?('windows')
+        package_absent(agent, "Ruby #{wins_ruby_version}")
+      else
+        package_absent(agent, 'ruby')
+        agent.add_env_var('PATH', beaker_path)
+      end
```

Comment:
Teardown block should be registered immediately after the path is modified, otherwise we may never get here, causing future tests to fail.

Reply:
Would that mean wrapping the above in a `teardown` block before  the other steps?

Reply:
Done